vmdb/app/controllers/api_controller/providers.rb

```diff
@@ -0,0 +1,158 @@
+class ApiController
+  module Providers
+    TYPE_ATTR         = "type"
+    CREDENTIALS_ATTR  = "credentials"
+    AUTH_TYPE_ATTR    = "auth_type"
+    DEFAULT_AUTH_TYPE = "default"
+    RESTRICTED_ATTRS  = [TYPE_ATTR, CREDENTIALS_ATTR] + %w(zone zone_id)
+
+    def create_resource_providers(type, _id, data = {})
+      if data.key?("id") || data.key?("href")
+        raise BadRequestError,
+              "Resource id or href should not be specified for creating a new #{type}"
+      end
+
+      create_provider(data)
+    end
+
+    def edit_resource_providers(type, id = nil, data = {})
+      raise BadRequestError, "Must specify an id for editing a #{type} resource" unless id
+      raise BadRequestError, "Provider type cannot be updated" if data.key?(TYPE_ATTR)
+
+      klass = collection_config[:providers][:klass].constantize
+      provider = resource_search(id, type, klass)
+      edit_provider(provider, data)
+    end
+
+    def refresh_resource_providers(type, id = nil, _data = nil)
+      raise BadRequestError, "Must specify an id for refreshing a #{type} resource" unless id
+
+      api_action(type, id) do |klass|
+        provider = resource_search(id, type, klass)
+        api_log_info("Refreshing #{provider_ident(provider)}")
+
+        refresh_provider(provider)
+      end
+    end
+
+    def delete_resource_providers(type, id = nil, _data = nil)
+      raise BadRequestError, "Must specify an id for deleting a #{type} resource" unless id
+
+      api_action(type, id) do |klass|
+        provider = resource_search(id, type, klass)
+        api_log_info("Deleting #{provider_ident(provider)}")
+
+        destroy_provider(provider)
+      end
+    end
+
+    private
+
+    def provider_ident(provider)
+      "Provider id:#{provider.id} name:'#{provider.name}'"
```

Comment:
do we want to implement `Provider#to_s` to provider so we don't need `provider_ident`?